Use PseudoElementIdentifier in ComputedStyleExtractor and CSSComputedStyleDeclaration

nt1m opened the pull request:
#### 741f7b16f8e63521bfd98aa8c0f81b99952b6d99
<pre>
Use PseudoElementIdentifier in ComputedStyleExtractor and CSSComputedStyleDeclaration
<a href="https://bugs.webkit.org/show_bug.cgi?id=268063">https://bugs.webkit.org/show_bug.cgi?id=268063</a>
<a href="https://rdar.apple.com/121580606">rdar://121580606</a>

Reviewed by Antti Koivisto.

In preparation of making animation code and `getComputedStyle()` work with pseudo-element name arguments (like `::highlight(name)` or `::view-transition-group(name)`).

Some of the consumers will also need refactoring to take PseudoElementIdentifier, which will be done separately.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getKeyframes):
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration):
(WebCore::CSSComputedStyleDeclaration::create):
(WebCore::CSSComputedStyleDeclaration::createEmpty):
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue const):
(WebCore::CSSComputedStyleDeclaration::copyProperties const):
(WebCore::CSSComputedStyleDeclaration::getPropertyValue const):
(WebCore::CSSComputedStyleDeclaration::length const):
(WebCore::CSSComputedStyleDeclaration::item const):
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue):
(WebCore::CSSComputedStyleDeclaration::getPropertyValue):
* Source/WebCore/css/CSSComputedStyleDeclaration.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::ComputedStyleExtractor):
(WebCore::ComputedStyleExtractor::getFontSizeCSSValuePreferringKeyword const):
(WebCore::ComputedStyleExtractor::useFixedFontDefaultSize const):
(WebCore::ComputedStyleExtractor::styledRenderer const):
(WebCore::computeRenderStyleForProperty):
(WebCore::ComputedStyleExtractor::customPropertyValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):
(WebCore::ComputedStyleExtractor::propertyMatches const):
(WebCore::ComputedStyleExtractor::getLayerCount const):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::styleForElementIgnoringPendingStylesheets):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::resolvePseudoElementStyle):
(WebCore::Element::computedStyle):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::computedStyle):
* Source/WebCore/dom/Node.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getComputedStyleForNode):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::getComputedStyle const):
(WebCore::LocalDOMWindow::getMatchedCSSRules const): Deleted.
(WebCore::LocalDOMWindow::webkitConvertPointFromNodeToPage const): Deleted.
(WebCore::LocalDOMWindow::webkitConvertPointFromPageToNode const): Deleted.
(WebCore::LocalDOMWindow::devicePixelRatio const): Deleted.
(WebCore::LocalDOMWindow::scrollBy const): Deleted.
(WebCore::LocalDOMWindow::scrollTo const): Deleted.
(WebCore::LocalDOMWindow::allowedToChangeWindowGeometry const): Deleted.
(WebCore::LocalDOMWindow::moveBy const): Deleted.
(WebCore::LocalDOMWindow::moveTo const): Deleted.
(WebCore::LocalDOMWindow::resizeBy const): Deleted.
(WebCore::LocalDOMWindow::resizeTo const): Deleted.
(WebCore::LocalDOMWindow::setTimeout): Deleted.
(WebCore::LocalDOMWindow::clearTimeout): Deleted.
(WebCore::LocalDOMWindow::setInterval): Deleted.
(WebCore::LocalDOMWindow::clearInterval): Deleted.
(WebCore::LocalDOMWindow::requestAnimationFrame): Deleted.
(WebCore::LocalDOMWindow::webkitRequestAnimationFrame): Deleted.
(WebCore::LocalDOMWindow::cancelAnimationFrame): Deleted.
(WebCore::LocalDOMWindow::requestIdleCallback): Deleted.
(WebCore::LocalDOMWindow::cancelIdleCallback): Deleted.
(WebCore::LocalDOMWindow::createImageBitmap): Deleted.
(WebCore::LocalDOMWindow::isSecureContext const): Deleted.
(WebCore::LocalDOMWindow::crossOriginIsolated const): Deleted.
(WebCore::didAddStorageEventListener): Deleted.
(WebCore::LocalDOMWindow::isSameSecurityOriginAsMainFrame const): Deleted.
(WebCore::LocalDOMWindow::addEventListener): Deleted.
(WebCore::LocalDOMWindow::deviceOrientationController const): Deleted.
(WebCore::LocalDOMWindow::deviceMotionController const): Deleted.
(WebCore::LocalDOMWindow::isAllowedToUseDeviceMotionOrOrientation const): Deleted.
(WebCore::LocalDOMWindow::isAllowedToUseDeviceMotion const): Deleted.
(WebCore::LocalDOMWindow::isAllowedToUseDeviceOrientation const): Deleted.
(WebCore::LocalDOMWindow::hasPermissionToReceiveDeviceMotionOrOrientationEvents const): Deleted.
(WebCore::LocalDOMWindow::startListeningForDeviceOrientationIfNecessary): Deleted.
(WebCore::LocalDOMWindow::stopListeningForDeviceOrientationIfNecessary): Deleted.
(WebCore::LocalDOMWindow::startListeningForDeviceMotionIfNecessary): Deleted.
(WebCore::LocalDOMWindow::stopListeningForDeviceMotionIfNecessary): Deleted.
(WebCore::LocalDOMWindow::failedToRegisterDeviceMotionEventListener): Deleted.
(WebCore::LocalDOMWindow::incrementScrollEventListenersCount): Deleted.
(WebCore::LocalDOMWindow::decrementScrollEventListenersCount): Deleted.
(WebCore::LocalDOMWindow::resetAllGeolocationPermission): Deleted.
(WebCore::LocalDOMWindow::removeEventListener): Deleted.
(WebCore::LocalDOMWindow::languagesChanged): Deleted.
(WebCore::LocalDOMWindow::dispatchLoadEvent): Deleted.
(WebCore::LocalDOMWindow::dispatchEvent): Deleted.
(WebCore::LocalDOMWindow::removeAllEventListeners): Deleted.
(WebCore::LocalDOMWindow::captureEvents): Deleted.
(WebCore::LocalDOMWindow::releaseEvents): Deleted.
(WebCore::LocalDOMWindow::finishedLoading): Deleted.
(WebCore::LocalDOMWindow::setLocation): Deleted.
(WebCore::LocalDOMWindow::printErrorMessage const): Deleted.
(WebCore::LocalDOMWindow::crossDomainAccessErrorMessage): Deleted.
(WebCore::LocalDOMWindow::isInsecureScriptAccess): Deleted.
(WebCore::LocalDOMWindow::createWindow): Deleted.
(WebCore::LocalDOMWindow::open): Deleted.
(WebCore::LocalDOMWindow::showModalDialog): Deleted.
(WebCore::LocalDOMWindow::enableSuddenTermination): Deleted.
(WebCore::LocalDOMWindow::disableSuddenTermination): Deleted.
(WebCore::LocalDOMWindow::frame const): Deleted.
(WebCore::LocalDOMWindow::protectedFrame const): Deleted.
(WebCore::LocalDOMWindow::eventListenersDidChange): Deleted.
(WebCore::LocalDOMWindow::cookieStore): Deleted.
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
* Source/WebCore/style/PseudoElementIdentifier.h:
* Source/WebCore/style/PseudoElementRequest.h:
(WebCore::Style::PseudoElementRequest::PseudoElementRequest):
(WebCore::Style::PseudoElementRequest::nameArgument const):
(WebCore::Style::PseudoElementRequest::pseudoElementNameArgument const): Deleted.
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::computedStyle):
* Source/WebCore/svg/SVGElement.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::computedStyleIncludingVisitedInfo const):

Canonical link: <a href="https://commits.webkit.org/273934@main">https://commits.webkit.org/273934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fff2c5918fc701b3e3f0b9168edfc438be7abf9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16151 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39546 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39812 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38553 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18731 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13295 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13621 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11863 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/33406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41072 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33724 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/33838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12196 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/35902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8411 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12547 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/style/PseudoElementRequest.h:
```diff
@@ -50,6 +51,13 @@ class PseudoElementRequest {
+        return !scrollbarState && pseudoElementNameArgument.isNull();
+    }
+
+    static PseudoElementRequest NoPseudoElement = { PseudoId::None };
```
Why is it `static` ?

nt1m replied:
It would be wasteful to re-define this for every instance of PseudoElementRequest, this was meant to be the default value for PseudoElementRequest (though it doesn't look like this compiles)

anttijk commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1692,7 +1692,11 @@ Ref<CSSStyleDeclaration> LocalDOMWindow::getComputedStyle(Element& element, cons
+        return CSSComputedStyleDeclaration::create(element, IsEmptyStyle::Yes);
+    if (*pseudoId == PseudoId::None)
+        return CSSComputedStyleDeclaration::create(element, std::nullopt);
+    return CSSComputedStyleDeclaration::create(element, Style::PseudoElementIdentifier { *pseudoId });
```
Factoring of this function is now bit odd and hard to follow for such a small function. How about separating the startsWith(":") path completely?

anttijk commented on Source/WebCore/css/CSSComputedStyleDeclaration.h:
```diff
@@ -33,11 +34,16 @@ namespace WebCore {
 class MutableStyleProperties;
 
+enum class AllowVisitedStyle : bool { No, Yes };
+enum class IsEmptyStyle : bool { No, Yes };
```
These could be scoped to CSSComputedStyleDeclaration and drop word 'Style'.

annevk replied:
If we expect callers to only ever pass Yes I think it would be nicer if the APIs are `createAllowVisited()` and `createEmpty()`.

anttijk commented on Source/WebCore/css/CSSComputedStyleDeclaration.h:
```diff
@@ -73,7 +80,8 @@ class CSSComputedStyleDeclaration final : public CSSStyleDeclaration, public Ref
     mutable Ref<Element> m_element;
-    std::optional<PseudoId> m_pseudoElementSpecifier { PseudoId::None };
+    std::optional<Style::PseudoElementIdentifier> m_pseudoElementIdentifier { std::nullopt };
+    bool m_isEmptyStyle { false };
```
'Style' doesn't add anything.

anttijk commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4107,46 +4108,53 @@ const RenderStyle* Element::resolveComputedStyle(ResolveComputedStyleMode mode)
-        style->setPseudoElementType(pseudoElementSpecifier);
+        style->setPseudoElementType(pseudoElementIdentifier.pseudoId);
+        if (!pseudoElementIdentifier.pseudoElementNameArgument.isNull())
+            style->setPseudoElementNameArgument(pseudoElementIdentifier.pseudoElementNameArgument);
```
RenderStyle should switch to PseudoElementIdentifier at some point.

nt1m replied:
Will file a bug

anttijk commented on Source/WebCore/style/PseudoElementIdentifier.h:
```diff
@@ -37,4 +37,9 @@ struct PseudoElementIdentifier {
 
+inline static bool isStyleCacheable(const PseudoElementIdentifier& identifier)
+{
+    return identifier.pseudoElementNameArgument.isNull();
```
`pseudoElementNameArgument` is an unnecessarily long name. How about just `argument`?

nt1m replied:
I'll change to `nameArgument`, some pseudo-elements have arguments that are not names (e.g. `::part()`)

anttijk commented on Source/WebCore/style/PseudoElementIdentifier.h:
```diff
@@ -37,4 +37,9 @@ struct PseudoElementIdentifier {
     AtomString pseudoElementNameArgument { nullAtom() };
 };
 
+inline static bool isStyleCacheable(const PseudoElementIdentifier& identifier)
```
Not sure this helper is helpful. Cacheable where? Also it does almost nothing.

nt1m replied:
Removed.

annevk commented on Source/WebCore/css/CSSComputedStyleDeclaration.h:
```diff
@@ -36,8 +37,12 @@ class MutableStyleProperties;
-    WEBCORE_EXPORT static Ref<CSSComputedStyleDeclaration> create(Element&, bool allowVisitedStyle);
-    static Ref<CSSComputedStyleDeclaration> create(Element&, std::optional<PseudoId>);
+    enum class AllowVisited : bool { No, Yes };
+    enum class IsEmpty : bool { No, Yes };
```
Nit: presumably this can be private now?

nt1m replied:
AllowVisited is still used outside, but yeah IsEmpty could be private.

annevk commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1690,9 +1690,13 @@ Ref<CSSStyleDeclaration> LocalDOMWindow::getComputedStyle(Element& element, cons
     std::optional<PseudoId> pseudoId = PseudoId::None;
     // FIXME: This does not work for pseudo-elements that take arguments (webkit.org/b/264103).
-    if (pseudoElt.startsWith(":"_s))
+    if (pseudoElt.startsWith(":"_s)) {
```
Nit: if you invert this you can have less nesting. And either way you can remove the upfront pseudoId declaration.

nt1m replied:
Good suggestion, done

darinadler commented on Source/WebCore/dom/Element.h:
```diff
@@ -30,6 +30,7 @@
 #include "EventOptions.h"
 #include "FocusOptions.h"
 #include "HitTestRequest.h"
+#include "PseudoElementIdentifier.h"
```
Don't need to add the same include to both `Node.h` and `Element.h`.

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -27,6 +27,7 @@
 #include "EventTarget.h"
 #include "ExceptionOr.h"
 #include "LayoutRect.h"
+#include "PseudoElementIdentifier.h"
```
It's not great to add a new include to a header as widely included as `Node.h`. It may be possible to rearrange things so that we only need a forward declaration here and if so, that would be good for build times. The vast majority of files including `Node.h` are not going be working with pseudo element identifiers. Often all we have to do is use overloading instead of default argument values to make this possible.

darinadler commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1688,11 +1688,14 @@ StyleMedia& LocalDOMWindow::styleMedia()
 Ref<CSSStyleDeclaration> LocalDOMWindow::getComputedStyle(Element& element, const String& pseudoElt) const
 {
-    std::optional<PseudoId> pseudoId = PseudoId::None;
+    if (!pseudoElt.startsWith(":"_s))
```
I believe there is a faster version of `startsWith` that takes a character rather than a string.

nt1m committed:
Use PseudoElementIdentifier in ComputedStyleExtractor and CSSComputedStyleDeclaration
https://bugs.webkit.org/show_bug.cgi?id=268063
rdar://121580606

Reviewed by Antti Koivisto.

In preparation of making animation code and `getComputedStyle()` work with pseudo-element name arguments (like `::highlight(name)` or `::view-transition-group(name)`).

Some of the consumers will also need refactoring to take PseudoElementIdentifier, which will be done separately.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getKeyframes):
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::CSSComputedStyleDeclaration):
(WebCore::CSSComputedStyleDeclaration::create):
(WebCore::CSSComputedStyleDeclaration::createEmpty):
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue const):
(WebCore::CSSComputedStyleDeclaration::copyProperties const):
(WebCore::CSSComputedStyleDeclaration::getPropertyValue const):
(WebCore::CSSComputedStyleDeclaration::length const):
(WebCore::CSSComputedStyleDeclaration::item const):
(WebCore::CSSComputedStyleDeclaration::getPropertyCSSValue):
(WebCore::CSSComputedStyleDeclaration::getPropertyValue):
* Source/WebCore/css/CSSComputedStyleDeclaration.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::ComputedStyleExtractor):
(WebCore::ComputedStyleExtractor::getFontSizeCSSValuePreferringKeyword const):
(WebCore::ComputedStyleExtractor::useFixedFontDefaultSize const):
(WebCore::ComputedStyleExtractor::styledRenderer const):
(WebCore::computeRenderStyleForProperty):
(WebCore::ComputedStyleExtractor::customPropertyValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):
(WebCore::ComputedStyleExtractor::propertyMatches const):
(WebCore::ComputedStyleExtractor::getLayerCount const):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::styleForElementIgnoringPendingStylesheets):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::resolvePseudoElementStyle):
(WebCore::Element::computedStyle):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::computedStyle):
* Source/WebCore/dom/Node.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getComputedStyleForNode):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::getComputedStyle const):
(WebCore::LocalDOMWindow::getMatchedCSSRules const): Deleted.
(WebCore::LocalDOMWindow::webkitConvertPointFromNodeToPage const): Deleted.
(WebCore::LocalDOMWindow::webkitConvertPointFromPageToNode const): Deleted.
(WebCore::LocalDOMWindow::devicePixelRatio const): Deleted.
(WebCore::LocalDOMWindow::scrollBy const): Deleted.
(WebCore::LocalDOMWindow::scrollTo const): Deleted.
(WebCore::LocalDOMWindow::allowedToChangeWindowGeometry const): Deleted.
(WebCore::LocalDOMWindow::moveBy const): Deleted.
(WebCore::LocalDOMWindow::moveTo const): Deleted.
(WebCore::LocalDOMWindow::resizeBy const): Deleted.
(WebCore::LocalDOMWindow::resizeTo const): Deleted.
(WebCore::LocalDOMWindow::setTimeout): Deleted.
(WebCore::LocalDOMWindow::clearTimeout): Deleted.
(WebCore::LocalDOMWindow::setInterval): Deleted.
(WebCore::LocalDOMWindow::clearInterval): Deleted.
(WebCore::LocalDOMWindow::requestAnimationFrame): Deleted.
(WebCore::LocalDOMWindow::webkitRequestAnimationFrame): Deleted.
(WebCore::LocalDOMWindow::cancelAnimationFrame): Deleted.
(WebCore::LocalDOMWindow::requestIdleCallback): Deleted.
(WebCore::LocalDOMWindow::cancelIdleCallback): Deleted.
(WebCore::LocalDOMWindow::createImageBitmap): Deleted.
(WebCore::LocalDOMWindow::isSecureContext const): Deleted.
(WebCore::LocalDOMWindow::crossOriginIsolated const): Deleted.
(WebCore::didAddStorageEventListener): Deleted.
(WebCore::LocalDOMWindow::isSameSecurityOriginAsMainFrame const): Deleted.
(WebCore::LocalDOMWindow::addEventListener): Deleted.
(WebCore::LocalDOMWindow::deviceOrientationController const): Deleted.
(WebCore::LocalDOMWindow::deviceMotionController const): Deleted.
(WebCore::LocalDOMWindow::isAllowedToUseDeviceMotionOrOrientation const): Deleted.
(WebCore::LocalDOMWindow::isAllowedToUseDeviceMotion const): Deleted.
(WebCore::LocalDOMWindow::isAllowedToUseDeviceOrientation const): Deleted.
(WebCore::LocalDOMWindow::hasPermissionToReceiveDeviceMotionOrOrientationEvents const): Deleted.
(WebCore::LocalDOMWindow::startListeningForDeviceOrientationIfNecessary): Deleted.
(WebCore::LocalDOMWindow::stopListeningForDeviceOrientationIfNecessary): Deleted.
(WebCore::LocalDOMWindow::startListeningForDeviceMotionIfNecessary): Deleted.
(WebCore::LocalDOMWindow::stopListeningForDeviceMotionIfNecessary): Deleted.
(WebCore::LocalDOMWindow::failedToRegisterDeviceMotionEventListener): Deleted.
(WebCore::LocalDOMWindow::incrementScrollEventListenersCount): Deleted.
(WebCore::LocalDOMWindow::decrementScrollEventListenersCount): Deleted.
(WebCore::LocalDOMWindow::resetAllGeolocationPermission): Deleted.
(WebCore::LocalDOMWindow::removeEventListener): Deleted.
(WebCore::LocalDOMWindow::languagesChanged): Deleted.
(WebCore::LocalDOMWindow::dispatchLoadEvent): Deleted.
(WebCore::LocalDOMWindow::dispatchEvent): Deleted.
(WebCore::LocalDOMWindow::removeAllEventListeners): Deleted.
(WebCore::LocalDOMWindow::captureEvents): Deleted.
(WebCore::LocalDOMWindow::releaseEvents): Deleted.
(WebCore::LocalDOMWindow::finishedLoading): Deleted.
(WebCore::LocalDOMWindow::setLocation): Deleted.
(WebCore::LocalDOMWindow::printErrorMessage const): Deleted.
(WebCore::LocalDOMWindow::crossDomainAccessErrorMessage): Deleted.
(WebCore::LocalDOMWindow::isInsecureScriptAccess): Deleted.
(WebCore::LocalDOMWindow::createWindow): Deleted.
(WebCore::LocalDOMWindow::open): Deleted.
(WebCore::LocalDOMWindow::showModalDialog): Deleted.
(WebCore::LocalDOMWindow::enableSuddenTermination): Deleted.
(WebCore::LocalDOMWindow::disableSuddenTermination): Deleted.
(WebCore::LocalDOMWindow::frame const): Deleted.
(WebCore::LocalDOMWindow::protectedFrame const): Deleted.
(WebCore::LocalDOMWindow::eventListenersDidChange): Deleted.
(WebCore::LocalDOMWindow::cookieStore): Deleted.
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::ruleMatches):
* Source/WebCore/style/PseudoElementIdentifier.h:
* Source/WebCore/style/PseudoElementRequest.h:
(WebCore::Style::PseudoElementRequest::PseudoElementRequest):
(WebCore::Style::PseudoElementRequest::nameArgument const):
(WebCore::Style::PseudoElementRequest::pseudoElementNameArgument const): Deleted.
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForPseudoElement):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::computedStyle):
* Source/WebCore/svg/SVGElement.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::computedStyleIncludingVisitedInfo const):

Canonical link: https://commits.webkit.org/273934@main
